Issue 1274: Unblock writer thread and log on all exceptional cases

shiveshr opened the pull request:
Note: recreating a new PR as the original PR is on a fork that no longer exist. 
(https://github.com/pravega/pravega/pull/1275)
I have updated this new PR by taking care of comments in the above PR.
 Change log description
Right now one thread is blocked while write.get() is called. We should unblock the thread using AckFuture.addListener
Also, log all exceptions on AckFuture.get() and not just executionException

Purpose of the change
Dont block on write.get and expand logging to cover all exceptions.
What the code does
does not block on ackfuture anymore
How to verify it
All existing unit tests should pass.
Also ran all end to end scale integration tests

fpj commented:
Wait, there is a pull request for this already: #1275, why are we creating another PR?

shiveshr commented:
@fpj i had to create a new branch as i am on a new fork and that branch was on my previous fork (shiveshr/pravega). Because of public/private access issue i ended up creating a new fork (shiveshr/pravega-1) so i created a new PR but i have referenced other PR because i have addressed comments from that PR and have also replied to those comments.

tkaitchuck commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java:
```diff
@@ -178,19 +179,37 @@ private void triggerScaleDown(String streamSegmentName, boolean silent) {
 
     private CompletableFuture<Void> writeRequest(ScaleEvent event) {
+        if (writer.get() == null) {
+            log.warn("Event writer not set up yet. Ignoring the request");
```
This is odd, why would this occur? If it shouldn't happen just make this a precondition.

shiveshr replied:
it shouldnt happen, but in the previous PR you gave a comment about null check. i will make this assert.

tkaitchuck replied:
I don't see an update

tkaitchuck commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java:
```diff
@@ -178,19 +179,37 @@ private void triggerScaleDown(String streamSegmentName, boolean silent) {
-                } catch (InterruptedException | ExecutionException e) {
-                    log.error("error sending request to requeststream {}", e);
+            AckFuture ackFuture = writer.get().writeEvent(event.getKey(), event);
+            ackFuture.addListener(() -> writeComplete.complete(null), executor);
```
This is actually incorrect. And I see why AckFuture is a bad interface... So hopefully we can get a replacement soon but get() still needs to be called inside the runnable as it could have failed.

shiveshr replied:
what is incorrect though?

tkaitchuck replied:
Yes, once #1356 goes in that insanity won't be necessary.

tkaitchuck commented on segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java:
```diff
@@ -178,19 +179,37 @@ private void triggerScaleDown(String streamSegmentName, boolean silent) {
+            AckFuture ackFuture = writer.get().writeEvent(event.getKey(), event);
+            ackFuture.addListener(() -> writeComplete.complete(null), executor);
+
+            writeComplete.thenAcceptAsync((Void v) -> {
```
This can be combined with the line above. This block can be the runnable passed to the listener.

shiveshr replied:
no, in the runnable we simply complete this future. 
writeComplete.complete(null)
So where we chain to this future (writeComplete) has nothing to do with the runnable we pass.

tkaitchuck replied:
Right, so you could if you wanted, remove writeComplete and run the logic below inside of the listener callback.

shiveshr commented:
Closing this PR. 
Depends on #1356 as ackfuture will be replaced post that change. So not merging this right now and will handle the chaining/transforms on completable future which will be returned post #1356 merge
